x-pack/plugins/maps/public/classes/sources/mvt_single_layer_vector_source/mvt_field_config_editor.tsx

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+/* eslint-disable @typescript-eslint/consistent-type-definitions */
+
+import React, { ChangeEvent, Component, Fragment } from 'react';
+import {
+  EuiButtonIcon,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiButtonEmpty,
+  EuiSuperSelect,
+  EuiFieldText,
+  EuiSpacer,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import _ from 'lodash';
+import { MVTFieldDescriptor } from '../../../../common/descriptor_types';
+import { FieldIcon } from '../../../../../../../src/plugins/kibana_react/public';
+import { MVT_FIELD_TYPE } from '../../../../common/constants';
+
+function makeOption({
+  value,
+  icon,
+  message,
+}: {
+  value: MVT_FIELD_TYPE;
+  icon: string;
+  message: string;
+}) {
+  return {
+    value,
+    inputDisplay: (
+      <EuiFlexGroup gutterSize="s" alignItems="center">
+        <EuiFlexItem grow={null}>
+          <FieldIcon type={icon} fill="none" />
+        </EuiFlexItem>
+        <EuiFlexItem>{message}</EuiFlexItem>
+      </EuiFlexGroup>
+    ),
+  };
+}
+
+const FIELD_TYPE_OPTIONS = [
+  {
+    value: MVT_FIELD_TYPE.STRING,
+    icon: 'string',
+    message: i18n.translate('xpack.maps.mvtSource.stringFieldLabel', {
+      defaultMessage: 'string',
+    }),
+  },
+  {
+    value: MVT_FIELD_TYPE.NUMBER,
+    icon: 'number',
+    message: i18n.translate('xpack.maps.mvtSource.numberFieldLabel', {
+      defaultMessage: 'number',
+    }),
+  },
+].map(makeOption);
+
+export interface Props {
+  fields: MVTFieldDescriptor[];
+  onChange: (fields: MVTFieldDescriptor[]) => void;
+}
+
+interface State {
+  currentFields: MVTFieldDescriptor[];
+}
+
+export class MVTFieldConfigEditor extends Component<Props, State> {
+  state: State = {
+    currentFields: _.cloneDeep(this.props.fields),
```

Comment:
We ideally shouldn't be initializing state with props this way, it's considered an anti-pattern. Maybe replace this logic with a constructor that does the same thing:

```
constructor(props) {
    super(props);
    this.state = _.cloneDeep(props.fields);
}
```

Reply:
Link? I believe both are equivalent. The current way is just JS-sugar for creating an instance variable, similar to how you'd create one at the start of constructor. But maybe this is different in React?

Reply:
I do not think there is a difference in react. I could also not find anything in kibana style guide about it https://github.com/elastic/kibana/blob/master/STYLEGUIDE.md.

Here is a stackoverflow question about this https://stackoverflow.com/questions/45451141/what-is-the-difference-between-using-constructor-vs-state-to-declare-state.

"
They are roughly equivalent. The significant difference is that the initializer in the second example is executed before constructor.

The second approach uses class fields proposal.
"

Before "class fields" props was not available outside of the constructor so you had to use the constructor. That is not the case with "class fields".

Personally, I like the simplicity of setting state without the noise of a constructor.

Reply:
My mistake, disregard then!